src/charm.py

```diff
@@ -0,0 +1,257 @@
+
+LibraryWriteBehaviour: typing.TypeAlias = typing.Literal['skip', 'clean']
+LIBRARY_WRITE_BEHAVIOURS = typing.get_args(LibraryWriteBehaviour)
+
+
+CalibreWebLayerDict = typing.TypedDict(
+    'CalibreWebLayerDict',
+    {
+        'summary': str,
+        'description': str,
+        'services': dict[str, ops.pebble.ServiceDict],
+    },
+    total=True,
+)
+
+
+class CalibreWebCharm(ops.CharmBase):
+    """Charm the calibre-web application."""
+
+    def __init__(self, framework: ops.Framework) -> None:
+        super().__init__(framework)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        framework.observe(self.on[CONTAINER_NAME].pebble_ready, self._on_pebble_ready)
+        framework.observe(self.on[STORAGE_NAME].storage_attached, self._on_storage_attached)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on[LIBRARY_WRITE_ACTION].action, self._on_library_write)
+        framework.observe(self.on[LIBRARY_INFO_ACTION].action, self._on_library_info)
+        self.unit.set_ports(8083)
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        logger.debug("_on_collect_status")
+        event.add_status(ops.ActiveStatus())
+        _, status = self._get_library_write_behaviour()
+        event.add_status(status)
+
+    def _on_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
+        """Define and start the workload using the Pebble API."""
+        container = event.workload
+        container.add_layer(SERVICE_NAME, {**self.get_pebble_layer()}, combine=True)
+        container.replan()
+
+    def _on_storage_attached(self, event: ops.StorageAttachedEvent) -> None:
+        self._push_library_to_storage()
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Don't do anything! User can run library-write action if needed.
+
+        Bad config values are handled in _on_collect_status
+        """
+        pass
+
+    def _on_library_write(self, event: ops.ActionEvent) -> None:
+        library_write_behaviour, status = self._get_library_write_behaviour()
+        if library_write_behaviour is None:
+            event.fail(status.message)
+            return
+        self._push_library_to_storage()
+        event.set_results({LIBRARY_WRITE_CONFIG: library_write_behaviour})
+
+    def _on_library_info(self, event: ops.ActionEvent) -> None:
+        container = self.framework.model.unit.containers[CONTAINER_NAME]
+        format = cast(str, event.params[LIBRARY_INFO_FORMAT_PARAM])
+        match format:
+            case 'tree':
+                try:
+                    container.exec(['which', 'tree']).wait()
+                except ops.pebble.ExecError:
+                    logger.debug('_on_library_info: tree: installing dependencies')
+                    container.exec(['apt', 'update']).wait()
+                    container.exec(['apt', 'install', 'tree', '-y']).wait()
```

Comment:
I find that a little weird.
You're ultimately in control of the docker image, so I would expect `tree` to be there from the start.

Reply:
You're right. If I wrote a new `Dockerfile` for `calibre-web` in the future then I would include it. My current flimsy justification for `tree` is that it's only needed for this case of this action, so it's better to only install it if the user invokes it, but  I don't think that really holds up. Also there's another case where I do this for something that will be needed on every deployment ...